File: 2020/02/main.zig
const std = @import("std");
const warn = std.debug.warn;

const input = @embedFile("input.txt");

fn validate1(min: u32, max: u32, char: u8, password: []const u8) bool {
    var count: u32 = 0;
    for (password) |c| {
        if (c == char) {
            count += 1;
        }
    }
    return count >= min and count <= max;
}

fn validate2(x: u32, y: u32, char: u8, password: []const u8) bool {
    return (password[x - 1] == char) != (password[y - 1] == char);
}

pub fn main() !void {
    var answer1: i32 = 0;
    var answer2: i32 = 0;
    var lines = std.mem.tokenize(input, "\n");
    while (lines.next()) |line| {
        var values = std.mem.tokenize(line, "- :");
        var x = try std.fmt.parseInt(u32, values.next().?, 10);
        var y = try std.fmt.parseInt(u32, values.next().?, 10);
        var char = values.next().?[0];
        var password = values.next().?;
        if (validate1(x, y, char, password)) answer1 += 1;
        if (validate2(x, y, char, password)) answer2 += 1;
    }
    warn("part 1: {}\n", .{answer1});
    warn("part 2: {}\n", .{answer2});
}

File: 2020/02/main.rs
fn validate_1(min: usize, max: usize, c: char, password: &str) -> bool {
	let mut count: usize = 0;
	for ch in password.chars() {
		if ch == c {
			count += 1;
		}
	}
	return count >= min && count <= max;
}

fn validate_2(x: usize, y: usize, c: char, password: &str) -> bool {
	let password: Vec<char> = password.chars().collect();
	return (password[x - 1] == c) != (password[y - 1] == c);
}

fn main() {
	let input = std::fs::read_to_string("2020/02/input.txt").unwrap();
	let mut answer_1: i32 = 0;
	let mut answer_2: i32 = 0;
	for line in input.lines() {
		let mut values = line.split(['-', ' '].as_ref());
		let x = values.next().unwrap().parse::<usize>().unwrap();
		let y = values.next().unwrap().parse::<usize>().unwrap();
		let c = values.next().unwrap().chars().next().unwrap();
		let password = values.next().unwrap();
		if validate_1(x, y, c, password) {
			answer_1 += 1;
		}
		if validate_2(x, y, c, password) {
			answer_2 += 1;
		}
	}
	println!("part 1: {}", answer_1);
	println!("part 2: {}", answer_2);
}
